### PR TITLE
DOC: Build docs with warnings as errors

### PR DESCRIPTION
**Pull request by larsoner**
Inspired by @rgommers in #9618 I'm trying to build with `-nWT --keep-going` (nitpicky, turn warnings into errors, full traceback on error, and keep going until the end to show an error). I now have it down to a few dozen lines of errors, then I'll enable "don't allow warnings in examples" (with exceptions of course) because it seems like good practice.

### CODE REVIEW

**Comment by rgommers** [comment]
This sounds awesome! If we can get it completely clean, we could add a check to CI to keep it that way.

**Comment by ilayn** [comment]
My inner OCD thanks you :)

**Comment by WarrenWeckesser** [comment]
@larsoner and @rgommers: thanks, it is great to see the docs getting some clean up.

@rgommers  wrote

>  If we can get it completely clean, we could add a check to CI to keep it that way.

This.  Without CI to ensure that it stays clean, fixing warnings in the docs is a Sisyphean task.

**Comment by larsoner** [comment]
Agreed. This PR will make the `make html-scipyorg` target throw an error by building with `-nWT`, and you can see it works/errors already here on CircleCI since I haven't fixed everything - true TDD! From what I've seen doing this on other repos, the errors produced by `-nWT` are pretty clear, too.

**Comment by larsoner** [comment]
Okay the CircleCI doc build is [now green](https://circleci.com/gh/scipy/scipy/10726).

Ready for review/merge from my end. Sorry for the huge diff, turns out our docs had a lot of little quirks to sort out.

File: doc/source/scipyoptdoc.py [hunk 1]
```diff
@@ -82,12 +83,17 @@ def run(self):
             # Interface function
             name = self.arguments[0].strip()
             obj = _import_object(name)
-            args, varargs, keywords, defaults = inspect.getargspec(obj)
+            # XXX deprecation that we should fix eventually
+            with warnings.catch_warnings(record=True):
+                warnings.simplefilter('ignore')
+                args, varargs, keywords, defaults = inspect.getargspec(obj)
```
Comment (ev-br) [inline doc/source/scipyoptdoc.py]: Is the deprecarion in inspect.getargspec? There's a shim in https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L249

Comment (larsoner) [reply doc/source/scipyoptdoc.py]: Two of the three are, yeah, thanks. I've replaced those.

File: doc/source/_templates/autosummary/class.rst [hunk 2]
```diff
@@ -1,27 +1,34 @@
...
-{% if methods %}
+.. currentmodule:: {{ module }}
+
+.. autoclass:: {{ objname }}
+   :members:
+   :special-members: __init__,__call__,__mul__,__getitem__,__len__
+
+  {% block methods %}
+  {% if methods %}
    .. HACK -- the point here is that we don't want this to appear in the output, but the autosummary should still generate the pages.
```
[the remark this answers is not in the document]
Comment (rgommers) [reply doc/source/_templates/autosummary/class.rst]: I guess comparing rendered html for a couple of big classes and call it good.

**Comment by rgommers** [comment]
I read through the first third of this PR, that's all I'll be able to do. Looks awesome though! I'd say verify the class rendering hasn't deteriorated, and then merge quickly - I think otherwise new issues will slip in.

Thanks so much @larsoner

**Comment by larsoner** [comment]
Good call -- `interp1d` on [1.2.0](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) and [this PR](https://10733-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.interpolate.interp1d.html) look different -- the methods are not shown on the page for the class, only via a link to a new page :(

I'll look into it to see if I can fix it.

**Comment by larsoner** [comment]
Actually, I think I prefer the new way, that is having all the methods listed on the same page as the class itself. But if others prefer the old way, I can change it so that separate pages are linked to for each method. Thoughts?

**Comment by larsoner** [comment]
For the sake of comparison

- `interp1d` has one method; docstring shown in [1.2.0](http://scipy.github.io/devdocs/generated/scipy.interpolate.interp1d.html) and [this PR](https://10754-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.interpolate.interp1d.html)
- `Rotation` has many methods; docstring shown in [1.2.0](http://scipy.github.io/devdocs/generated/scipy.spatial.transform.Rotation.html) and [this PR](https://10754-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.spatial.transform.Rotation.html).

**Comment by larsoner** [comment]
On second thought, I reverted the change to have all the methods on the same page. I think the `HACK` comment that was in `class.rst` was designed to have them on separate pages, so we'll keep it that way. Latest commit does this, and fixes some challenging broken links having to do with inherited methods.

In any case, here are the updated ones, which look good to me:

- `interp1d` has one method; docstring shown in [1.2.0](http://scipy.github.io/devdocs/generated/scipy.interpolate.interp1d.html) and [this PR](https://10817-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.interpolate.interp1d.html) are almost identical, except the `fill_value` link is fixed (and `dtype` is removed from the list since it's not documented / useful)
- `Rotation` has many methods; docstring shown in [1.2.0](http://scipy.github.io/devdocs/generated/scipy.spatial.transform.Rotation.html) and [this PR](https://10817-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.spatial.transform.Rotation.html) look identical
- `CloughTocher2DInterpolator` in [1.2.0](http://scipy.github.io/devdocs/generated/scipy.interpolate.CloughTocher2DInterpolator.html) has a broken link to `__call__` which is fixed [in this PR](https://10817-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.interpolate.CloughTocher2DInterpolator.html)

**Comment by larsoner** [comment]
Rebased to check to see if there were any new errors in the last week and there were a couple. I fixed them and push-forced.

**Comment by ilayn** [comment]
pypy3 failure is not related. A very happy +1 for merge from me.

**Comment by ev-br** [comment]
OK, great, let's get this in. Thanks Eric, all.

**Pull request merged**